pkg/steps/source.go

```diff
@@ -539,20 +542,51 @@ func hintsAtInfraReason(logSnippet string) bool {
 		strings.Contains(logSnippet, "connection reset by peer")
 }
 
-func waitForBuildOrTimeout(ctx context.Context, buildClient BuildClient, namespace, name string) error {
-	return waitForBuild(ctx, buildClient, namespace, name)
+func waitForBuildOrTimeout(
+	ctx context.Context,
+	buildClient BuildClient,
+	podClient kubernetes.PodClient,
+	namespace, name string,
+) error {
+	return waitForBuild(ctx, buildClient, podClient, namespace, name)
 }
 
-func waitForBuild(ctx context.Context, buildClient BuildClient, namespace, name string) error {
+func waitForBuild(
+	ctx context.Context,
+	buildClient BuildClient,
+	podClient kubernetes.PodClient,
+	namespace, name string,
+) error {
 	logrus.WithFields(logrus.Fields{
 		"namespace": namespace,
 		"name":      name,
 	}).Trace("Waiting for build to be complete.")
-
-	evaluatorFunc := func(obj runtime.Object) (bool, error) {
-		switch build := obj.(type) {
-		case *buildapi.Build:
+	var ret atomic.Pointer[buildapi.Build]
+	var eg *errgroup.Group
+	eg, ctx = errgroup.WithContext(ctx)
+	pendingCtx, cancel := context.WithCancel(ctx)
+	pendingCheck := func() error {
+		timeout := podClient.GetPendingTimeout()
+		select {
+		case <-pendingCtx.Done():
+		case <-time.After(time.Until(ret.Load().CreationTimestamp.Add(timeout))):
+			if err := checkPending(ctx, podClient, ret.Load(), timeout, time.Now()); err != nil {
+				logrus.Infof(err.Error())
+				return err
+			}
+		}
+		return nil
+	}
+	eg.Go(func() error {
+		defer cancel()
+		return kubernetes.WaitForConditionOnObject(ctx, buildClient, ctrlruntimeclient.ObjectKey{Namespace: namespace, Name: name}, &buildapi.BuildList{}, &buildapi.Build{}, func(obj runtime.Object) (bool, error) {
+			build := obj.(*buildapi.Build)
+			first := ret.Swap(build) == nil
 			switch build.Status.Phase {
```

Comment:
All right this is more a vague consideration rather than anything else. The [--pod-pending-timeout](https://github.com/openshift/ci-tools/blob/09cdfa1aefc4a7d11a44e148f362fede533b5689/cmd/ci-operator/main.go#L434) flag states that it controls the:
> "Maximum amount of time created containers can spend before the running state"  

but this switch does not even mention the state `buildapi.BuildPhaseRunning` in any place.
This obviously means nothing, but I feel like this logic is a little bit hidden.

Reply:
I agree the argument's description is not ideal, updated.